test/output-app/model/Tweet.js

```diff
@@ -1,60 +1,115 @@
 
 export default class Tweet {
   constructor(context) {
     this.context = context;
     this.collection = context.db.collection('tweet');
     this.pubsub = context.pubsub;
-    this.loader = new DataLoader(ids => findByIds(this.collection, ids));
+    let authQuery;
+    try {
+      const { me, User } = context;
+      authQuery = queryForRoles(me, ['admin', 'world'], ['authorId', 'coauthorsIds'], { User }, authlog('tweet findOneById', 'readOne', me));
+    } catch (err) { 
+      log.error(err.message);
+      authQuery = {_id: false}; // otherwise admin access
+    }
+    this.authorizedLoader = new DataLoader(ids => findByIds(this.collection, ids, authQuery));
   }
 
-  findOneById(id) {
-    return this.loader.load(id);
+  async findOneById(id, me, resolver) {
+    try {
+      return await this.authorizedLoader.load(id);
+    } catch (err) { log.error(err.message); }
   }
 
-  all({ lastCreatedAt = 0, limit = 10 }) {
-    return this.collection.find({
-      createdAt: { $gt: lastCreatedAt },
-    }).sort({ createdAt: 1 }).limit(limit).toArray();
+  find({ lastCreatedAt = 0, limit = 10, baseQuery = { createdAt: { $gt: lastCreatedAt } } }, me, resolver) {
+    try {
+      const authQuery = queryForRoles(me, ['admin', 'world'], ['authorId', 'coauthorsIds'], { User: this.context.User }, authlog(resolver, 'readMany', me));
+      const finalQuery = {...baseQuery, ...authQuery};
+      return this.collection.find(finalQuery).sort({ createdAt: 1 }).limit(limit).toArray();
+    } catch (err){ log.error(err.message); }
   }
 
-  author(tweet) {
-    return this.context.User.findOneById(tweet.authorId);
+  createdBy(tweet, me, resolver) {
+    return this.context.User.findOneById(tweet.createdById, me, resolver);
   }
 
-  likers(tweet, { lastCreatedAt = 0, limit = 10 }) {
-    return this.context.User.collection.find({
-      likedIds: tweet._id,
-      createdAt: { $gt: lastCreatedAt },
-    }).sort({ createdAt: 1 }).limit(limit).toArray();
+  updatedBy(tweet, me, resolver) {
+    return this.context.User.findOneById(tweet.updatedById, me, resolver);
   }
 
-  async insert(doc) {
-    const docToInsert = Object.assign({}, doc, {
-      createdAt: Date.now(),
-      updatedAt: Date.now(),
-    });
-    const id = (await this.collection.insertOne(docToInsert)).insertedId;
-    this.pubsub.publish('tweetInserted', await this.findOneById(id));
-    return id;
+  author(tweet, me, resolver) {
+    return this.context.User.findOneById(tweet.authorId, me, resolver);
   }
 
-  async updateById(id, doc) {
-    const ret = await this.collection.update({ _id: id }, {
-      $set: Object.assign({}, doc, {
-        updatedAt: Date.now(),
-      }),
-    });
-    this.loader.clear(id);
-    this.pubsub.publish('tweetUpdated', await this.findOneById(id));
-    return ret;
+  coauthors(tweet, { lastCreatedAt = 0, limit = 10 }, me, resolver) {
+    const baseQuery = {_id: { $in: tweet.coauthorsIds }, createdAt: { $gt: lastCreatedAt } };
```

Comment:
We won't need to set the `createdAt` bit here if we are a bit more careful above.